phoenix-core-client/src/main/java/org/apache/phoenix/expression/RowValueConstructorExpression.java

```diff
@@ -220,14 +223,24 @@ public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
                             // we'd have an ambiguity if this value happened to be the
                             // min possible value.
                             previousCarryOver = childType == null || childType.isFixedWidth();
-                            int bytesToWrite = getExpressionByteCount(child);
-                            for (int m = 0; m < bytesToWrite; m++) {
-                                output.write(QueryConstants.SEPARATOR_BYTE);
+                            if (childType == PVarbinaryEncoded.INSTANCE) {
+                                output.write(QueryConstants.VARBINARY_ENCODED_SEPARATOR_BYTES);
+                            } else {
+                                int bytesToWrite = getExpressionByteCount(child);
+                                for (int m = 0; m < bytesToWrite; m++) {
+                                    output.write(QueryConstants.SEPARATOR_BYTE);
+                                }
                             }
                         } else {
                             output.write(tempPtr.get(), tempPtr.getOffset(), tempPtr.getLength());
                             if (!childType.isFixedWidth()) {
-                                output.write(SchemaUtil.getSeparatorByte(true, false, child));
+                                if (childType != PVarbinaryEncoded.INSTANCE) {
+                                    output.write(SchemaUtil.getSeparatorByte(true, false, child));
+                                } else {
+                                    output.write(
+                                        SchemaUtil.getSeparatorBytesForVarBinaryEncoded(true, false,
+                                            child.getSortOrder()));
+                                }
```

Comment:
Could we introduce an API which always returns a byte array and we simply iterate over the array. That should work regardless whether there is one separator byte or multiple and reduce the amount of code duplication.